src/inmanta/server/services/resourceservice.py

```diff
@@ -563,31 +612,46 @@ async def get_increment(
         :param version: The version of the model to consider.
         :param connection: connection to use towards the DB.
             When the connection is in a transaction, we will always invalidate the cache
+        :param run_ahead_lock: lock used to keep agents hanging while building up the latest version
         """
 
-        def _get_cache_entry() -> Optional[tuple[abc.Set[ResourceIdStr], abc.Set[ResourceIdStr]]]:
+        async def _get_cache_entry() -> (
+            Optional[tuple[int, abc.Set[ResourceIdStr], abc.Set[ResourceIdStr], abc.Mapping[str, abc.Set[ResourceIdStr]]]]
+        ):
             """
-            Returns a tuple (increment, negative_increment) if a cache entry exists for the given environment and version
+            Returns a tuple (increment, negative_increment, negative_increment_per_agent)
+            if a cache entry exists for the given environment and version
             or None if no such cache entry exists.
             """
             cache_entry = self._increment_cache.get(env.id, None)
             if cache_entry is None:
                 # No cache entry found
                 return None
-            (version_cache_entry, incr, neg_incr) = cache_entry
-            if version_cache_entry != version:
+            (version_cache_entry, incr, neg_incr, neg_incr_per_agent, run_ahead_lock) = cache_entry
+            if version_cache_entry > version:
+                # Cache is ahead,
+                if run_ahead_lock is not None:
+                    await run_ahead_lock.wait()
+            elif version_cache_entry != version:
                 # Cache entry exists for another version
+                # Expire
                 return None
-            return incr, neg_incr
+            return version_cache_entry, incr, neg_incr, neg_incr_per_agent
 
-        increment: Optional[tuple[abc.Set[ResourceIdStr], abc.Set[ResourceIdStr]]] = _get_cache_entry()
+        increment: Optional[
+            tuple[int, abc.Set[ResourceIdStr], abc.Set[ResourceIdStr], abc.Mapping[str, abc.Set[ResourceIdStr]]]
+        ] = await _get_cache_entry()
         if increment is None or (connection is not None and connection.is_in_transaction()):
             lock = self._increment_cache_locks[env.id]
             async with lock:
```

Comment:
Can't this lock create a deadlock with the `run_ahead_lock`? In the release flow the `run_ahead lock` is acquired first but if a fast agent calls `get_increment` before the release flow does it acquires this lock first. Then it starts waiting for the `run_ahead_lock` which will in turn start waiting for this lock.

Reply:
the release flow never acquires the `run_ahead_lock` , it creates it. 
- the release flow is always head, so it passes in the lock for others, but doesn't every block on it
- the run_ahead_lock is placed just prior to leaving the critical section here, so it will be resolved.

Reply:
Doesn't it acquire it in order to populate the associated cache? i.e. [here](https://github.com/inmanta/inmanta-core/pull/7278/files#diff-e3723a8f8dde22b64bfd475f55960927014a3896679d1ca244b250dd8b1e5987R1154). So it creates the lock, and then eventually populates the cache by calling into the same flow that acquires both this lock and the other one.

Reply:
Ah no, I see. The context manager does not place the lock, it's the `get_increment` call itself that places it. And indeed it is only placed after populating the cache (even if it is created before).